### DISCUSSION TITLE
Take advantage of Google's OSS-Fuzz testing for OSS projects

### DISCUSSION THREAD

**Discussion by jqnatividad** [ckan/ckan#6677] [Ideas]
CKAN has been identified as a critical open source project (https://github.com/ossf/criticality_score#public-data).

Perhaps, we should integrate Google's OSS-Fuzz service (https://google.github.io/oss-fuzz/) and have "continuous fuzzing" integrated into CKAN's CI testing (https://google.github.io/oss-fuzz/getting-started/continuous-integration/).

Google also offers integration rewards (https://google.github.io/oss-fuzz/getting-started/integration-rewards/).

**Comment by EricSoroos** [comment]
I've had some experience with integrating OSS-Fuzz on a python project. Apart from the general pain integrating python into fuzzing libraries in a performant manner, I think CKAN is going to be somewhat tricky to meaningfully integrate for a few reasons:

1) CKAN is a web application, effectively there's one input (the http stream) but interesting islands of behavior in that stream are going to be sparse.  More typical examples of items that are effective to fuzz would be media file decoders, where there are many byte changes that will alter the execution path. 
2) CKAN is heavily dependent on external state -- config files and data in solr and postgres. Assembling and referencing that state in the OSS fuzz environment is going to be a challenge.  
3) CKAN's startup speed will heavily impact the success of fuzzing. Getting 10 to 100 executions/second is a good baseline for getting results. 
4) CKAN doesn't have any code written in memory unsafe languages. OSS Fuzzing Python will generally get you memory access errors from extensions, slow unit tests, timeouts, and OOM issues. With no c extensions, CKAN will likely only find the timeouts and potentially a memory exhaustion issue.

**Reply by jqnatividad** [reply]
Thanks @EricSoroos for the insight...

What do you think about fuzz testing CKAN's API, given how CKAN's front-end extensively dogfoods the API and does just about everything through it?
